bpo-10746: ctypes: Fix PEP 3118 type codes for c_long, c_bool, c_int

pv opened the pull request:
Ctypes currently produces wrong pep3118 type codes for several types.
E.g. memoryview(ctypes.c_long()).format gives "<l" on 64-bit platforms,
but it should be "<q" instead for sizeof(c_long) == 8

The problem is that the '<>' endian specification in the struct syntax
also turns on the "standard size" mode, which makes type characters have
a platform-independent meaning, which does not match with the codes used
internally in ctypes.  The struct module format syntax also does not
allow specifying native-size non-native-endian items.

This commit adds a converter function that maps the internal ctypes
codes to appropriate struct module standard-size codes in the pep3118
format strings. The tests are modified to check for this.

Example of the current problem in practice:
```
>>> import numpy, ctypes
>>> numpy.asarray(memoryview(ctypes.c_long(42)))
/usr/lib64/python3.5/site-packages/numpy/core/numeric.py:482: RuntimeWarning: Item size computed from the PEP 3118 buffer format string does not match the actual item size.
  return array(a, dtype, copy=False, order=order)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.5/site-packages/numpy/core/numeric.py", line 482, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: setting an array element with a sequence.
```

- [struct format string syntax](https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment)
- [bpo-10746](https://bugs.python.org/issue10746)

<!-- issue-number: bpo-10746 -->
https://bugs.python.org/issue10746
<!-- /issue-number -->

SylvainCorlay commented:
http://bugs.python.org/issue26689

SylvainCorlay commented:
@dstufft any chance to get this one in? I added a link to the bugs.python.org issue.

brettcannon commented:
Thanks for the PR, but https://github.com/python/cpython is a cpython mirror repository; pull requests are not accepted on this repo. Please submit bugs and patches at https://bugs.python.org.

SylvainCorlay commented:
Thanks Brett. As you can see in the description of the PR, there is a link
to the patch on bugs.python.org which has not gotten much attention.
On Jul 11, 2016 2:08 PM, "Brett Cannon" <notifications@github.com> wrote:

> Closed #31 <https://github.com/python/cpython/pull/31>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#event-719093844>, or mute the
> thread
> <https://github.com/notifications/unsubscribe/ACSXFppvI6EG9nvXG6stNIOqy2KqLxUXks5qUoaCgaJpZM4H-cRy>
> .
>

pv committed:
bpo-10746: Fix ctypes PEP 3118 type codes for c_long, c_bool, c_int

Ctypes currently produces wrong pep3118 type codes for several types.
E.g. memoryview(ctypes.c_long()).format gives "<l" on 64-bit platforms,
but it should be "<q" instead for sizeof(c_long) == 8

The problem is that the '<>' endian specification in the struct syntax
also turns on the "standard size" mode, which makes type characters have
a platform-independent meaning, which does not match with the codes used
internally in ctypes.  The struct module format syntax also does not
allow specifying native-size non-native-endian items.

This commit adds a converter function that maps the internal ctypes
codes to appropriate struct module standard-size codes in the pep3118
format strings. The tests are modified to check for this.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow these steps to rectify the issue:

1. Sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/)
2. **Wait at least a day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
3. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

pv commented:
CLA signed.

pitrou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -249,6 +249,71 @@ PyDict_GetItemProxy(PyObject *dict, PyObject *key)
+  since the endianness may need to be swapped to a non-native one
+  later on.
+ */
+char *
```
This function should be declared `static`.

pitrou commented on Lib/ctypes/test/test_pep3118.py:
```diff
@@ -17,14 +17,47 @@ def normalize(format):
+        # Since some types may be aliases to each other, look up the
+        # struct code based on the actual ctypes type
+        tp = code_map[ctypes_code][0]
+        struct_code = code_map[tp._type_][1][sizeof(tp)]
```
I don't understand this line. Can you show an example where `tp._type_` is necessary?

pv replied:
I must say I don't understand it either anymore. It may be leftover from refactoring.
Since the tests seem to pass, I now simplified it.

pv replied:
Ah no, it was due to a int/long failure on windows:
```
======================================================================
FAIL: test_native_types (ctypes.test.test_pep3118.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\cpython\lib\ctypes\test\test_pep3118.py", line 56, in test_native_types
    normalize(replace_type_codes(fmt)))
AssertionError: '<l' != '<i'
- <l
+ <i
```
Added explanatory comment in the test.

pitrou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -249,6 +249,71 @@ PyDict_GetItemProxy(PyObject *dict, PyObject *key)
+    case 'i': pep_code = 'q'; break;
+    case 'I': pep_code = 'Q'; break;
+#else
+# error
```
It would be nice to add a non-empty error message, for example "SIZEOF_INT is not one of the expected values".

pitrou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -249,6 +249,71 @@ PyDict_GetItemProxy(PyObject *dict, PyObject *key)
+    case 'l': pep_code = 'q'; break;
+    case 'L': pep_code = 'Q'; break;
+#else
+# error
```
Ditto.

pitrou commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -249,6 +249,71 @@ PyDict_GetItemProxy(PyObject *dict, PyObject *key)
+#elif SIZEOF__BOOL == 8
+    case '?': pep_code = 'Q'; break;
+#else
+# error
```
Ditto.

pv commented:
Sorry for the delay, updated the PR based on comments.

pv commented:
ping?

pitrou commented:
Sorry for the delay.

pitrou commented on Lib/ctypes/test/test_pep3118.py:
```diff
@@ -17,14 +17,49 @@ def normalize(format):
+        #
+        # Example: on Windows c_long translates to '<i', not '<l'.
+        tp = code_map[ctypes_code][0]
+        struct_code = code_map[tp._type_][1][sizeof(tp)]
```
I still find this difficult to read and understand. Can't you rewrite the unit tests and the `native_types` array to get rid of the double indirection?

pitrou commented on Lib/ctypes/test/test_pep3118.py:
```diff
@@ -111,7 +146,11 @@ class Complete(Structure):
 #
 # This table contains format strings as they look on little endian
 # machines.  The test replaces '<' with '>' on big endian machines.
+# The type codes written below are the internal ctypes type codes,
```
Internal ctypes type codes are not even documented, are they?

pitrou commented:
Thanks for persisting. I am not really happy about the way this hacks around the current structure of the tests. It adds a lot of complication that seems pointless given the task at hand (we're just checking a mapping, right?).

pv commented:
A static translation table does not exist, because the size of the types,
and also which type code is produced for some of the integer types when
multiple equivalent possibilities exist, is platform and possibly compiler
dependent. If you are suggesting to write all possibilities explicitly, you
have to write different tables for win32, win64, linux32, etc. Can you
specify more precisely what you object to:
- using replace in strings to produce appropriate codes rather than writing
them explicitly,
- not hardcoding codes for equivalent integer types,
- something else?

26.8.2017 11.04 "Antoine Pitrou" <notifications@github.com> kirjoitti:

> Thanks for persisting. I am not really happy about the way this hacks
> around the current structure of the tests. It adds a lot of complication
> that seems pointless given the task at hand (we're just checking a mapping,
> right?).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#issuecomment-325105406>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5qKVqObam9UqEqb2IfRqEGfrycROks5sb9-AgaJpZM4L-Rci>
> .
>

pv commented:
The internal types codes are not documented. But do you have better
suggestions how to write the strings? They cannot be written in PEP3118
format because the produced strings are platform etc dependent.

26.8.2017 11.04 "Antoine Pitrou" <notifications@github.com> kirjoitti:

> Thanks for persisting. I am not really happy about the way this hacks
> around the current structure of the tests. It adds a lot of complication
> that seems pointless given the task at hand (we're just checking a mapping,
> right?).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#issuecomment-325105406>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5qKVqObam9UqEqb2IfRqEGfrycROks5sb9-AgaJpZM4L-Rci>
> .
>

pv commented:
This of course was ignored previously in the tests, because the ctypes
implementation of this stuff was just wrong.

26.8.2017 11.37 "Pauli Virtanen" <pav@iki.fi> kirjoitti:

The internal types codes are not documented. But do you have better
suggestions how to write the strings? They cannot be written in PEP3118
format because the produced strings are platform etc dependent.

26.8.2017 11.04 "Antoine Pitrou" <notifications@github.com> kirjoitti:

> Thanks for persisting. I am not really happy about the way this hacks
> around the current structure of the tests. It adds a lot of complication
> that seems pointless given the task at hand (we're just checking a mapping,
> right?).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#issuecomment-325105406>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5qKVqObam9UqEqb2IfRqEGfrycROks5sb9-AgaJpZM4L-Rci>
> .
>

pitrou commented:
My main gripe here is that you shouldn't need `code_map` at all, let alone the complicated double indirection inside that table. Just add/replace whatever information is missing to `native_types`.

Also we don't care at all about ctypes internal code types, so the tests would be clearer if they didn't involve them.

pv commented:
The expected pep strings are platform-dependent. Do you agree with that
statement? Do you mean the entries in the native_types table should be
translate_typecodes("...") calls in the table itself. Or, do you mean
code_map should be replaced by a slew of if statemebts encoding the
platform dependence?


26.8.2017 19.49 "Antoine Pitrou" <notifications@github.com> kirjoitti:

> My main gripe here is that you shouldn't need code_map at all, let alone
> the complicated double indirection inside that table. Just add/replace
> whatever information is missing to native_types.
>
> Also we don't care at all about ctypes internal code types, so the tests
> would be clearer if they didn't involve them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#issuecomment-325145915>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5i6dyAcsuUqZDCKqIIkcgKtizybgks5scEyagaJpZM4L-Rci>
> .
>

pv commented:
I think I need railroad track now :)
Could you give an example how you would like the test for the pep3118
format string for c_long to look like?


26.8.2017 20.08 "Pauli Virtanen" <pav@iki.fi> kirjoitti:

The expected pep strings are platform-dependent. Do you agree with that
statement? Do you mean the entries in the native_types table should be
translate_typecodes("...") calls in the table itself. Or, do you mean
code_map should be replaced by a slew of if statemebts encoding the
platform dependence?


26.8.2017 19.49 "Antoine Pitrou" <notifications@github.com> kirjoitti:

> My main gripe here is that you shouldn't need code_map at all, let alone
> the complicated double indirection inside that table. Just add/replace
> whatever information is missing to native_types.
>
> Also we don't care at all about ctypes internal code types, so the tests
> would be clearer if they didn't involve them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#issuecomment-325145915>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5i6dyAcsuUqZDCKqIIkcgKtizybgks5scEyagaJpZM4L-Rci>
> .
>

pitrou commented:
You are already encoding platform-dependent information (the size mappings) in `code_map`? Why don't you fold that into `native_types`?

pv commented:
native_types and the other tables contain also compound types where these
type codes appear. To me the present approach appeared the simplest solution
without code repetition.

26.8.2017 20.13 "Pauli Virtanen" <pav@iki.fi> kirjoitti:

> I think I need railroad track now :)
> Could you give an example how you would like the test for the pep3118
> format string for c_long to look like?
>
>
> 26.8.2017 20.08 "Pauli Virtanen" <pav@iki.fi> kirjoitti:
>
> The expected pep strings are platform-dependent. Do you agree with that
> statement? Do you mean the entries in the native_types table should be
> translate_typecodes("...") calls in the table itself. Or, do you mean
> code_map should be replaced by a slew of if statemebts encoding the
> platform dependence?
>
>
> 26.8.2017 19.49 "Antoine Pitrou" <notifications@github.com> kirjoitti:
>
>> My main gripe here is that you shouldn't need code_map at all, let alone
>> the complicated double indirection inside that table. Just add/replace
>> whatever information is missing to native_types.
>>
>> Also we don't care at all about ctypes internal code types, so the tests
>> would be clearer if they didn't involve them.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/python/cpython/pull/31#issuecomment-325145915>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AACI5i6dyAcsuUqZDCKqIIkcgKtizybgks5scEyagaJpZM4L-Rci>
>> .
>>
>
>

pv commented:
Ok, once more then, without magic. Is this what you wanted?

pitrou commented:
Thank you! :-) I find this much better. In particular, it's explicit that the exact type code does not depend only on the type size but also on whether `c_int` aliases `c_long`, for example (which is a weird oddity, but I guess it's not this PR's responsability to change that).

One additional thing: can you add a NEWS entry? We nowadays use the [blurb CLI tool](https://devguide.python.org/committing/?highlight=blurb#what-s-new-and-news-entries) for that.

pv commented:
Added news blurb.

pv commented:
Btw, the travis tests don't somehow seem to run, giving a strange message "fatal: Invalid symmetric difference expression a30f6d45ac3e72761b96a8df0527182029eaee24...b1eccfaff3fde214d261b5b64b589e3007a4e3bc
Only docs were updated, stopping build process." instead. I guess it's unrelated to the changes here, though --- it looks as if travis somehow didn't fetch the upstream master commit, so maybe it would just need `git fetch origin master` etc., before trying to do `git diff --name-only $TRAVIS_COMMIT_RANGE` in travis.yml

pitrou commented:
> Btw, the travis tests don't somehow seem to run, giving a strange message "fatal: Invalid symmetric difference expression

Oh, right. That means you should merge from master (or rebase if that's your thing).

pv commented:
Rebased, but it didn't seem to help: it says "Invalid symmetric difference expression a30f6d45ac3e72761b96a8df0527182029eaee24...4cb740948ecc6286efbcaac9b1fb1289d28080a4"
The version range is certainly correct, so I don't know what's going on there...

pitrou commented:
I don't see any such issue. Perhaps you just forgot to refresh this page?

pv commented:
It's here: https://travis-ci.org/python/cpython/jobs/269157668
The build passes with success regardless.

pitrou commented:
Hmm... are you sure you updated master before rebasing?

pv commented:
See the pushed commit. It's parent is a30f6d45, which is the current python
master.

28.8.2017 14.57 "Antoine Pitrou" <notifications@github.com> kirjoitti:

> Hmm... are you sure you updated master before rebasing?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/cpython/pull/31#issuecomment-325333792>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5tR7nhHuhdygAJlPE6t0pEI26cNrks5scqsxgaJpZM4L-Rci>
> .
>

pitrou commented:
I see. Well, it passed on AppVeyor anyway, so I'm going to merge. Thanks for persisting in this!
